src/bitwarden_fmt_plug.c

```diff
@@ -0,0 +1,215 @@
+#else
+#define ALGORITHM_NAME          "PBKDF2-SHA256 AES 32/" ARCH_BITS_STR " " SHA2_LIB
+#endif
+#endif
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        -1
+#define BINARY_SIZE             0
+#define PLAINTEXT_LENGTH        125
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define BINARY_ALIGN            1
+#define SALT_ALIGN              sizeof(uint64_t)
+#ifdef SIMD_COEF_32
+#define MIN_KEYS_PER_CRYPT      SSE_GROUP_SZ_SHA256
+#define MAX_KEYS_PER_CRYPT      (SSE_GROUP_SZ_SHA256 * 32)
+#else
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      256
+#endif
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int *cracked, cracked_count;
+static struct custom_salt *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+	saved_key = mem_calloc(sizeof(*saved_key),  self->params.max_keys_per_crypt);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(cracked);
+	MEM_FREE(saved_key);
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void bitwarden_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(*saved_key));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int bitwarden_decrypt(struct custom_salt *cur_salt, unsigned char *key)
+{
+	unsigned char ivec[IVLEN];
+	unsigned char out[BLOBLEN];
+	AES_KEY aes_decrypt_key;
+
+        AES_set_decrypt_key(key, 256, &aes_decrypt_key);
+        // memcpy(ivec, cur_salt->iv, 16);
+        AES_cbc_encrypt(cur_salt->blob + BLOBLEN - 32, out, BLOBLEN, &aes_decrypt_key, ivec, AES_DECRYPT);
+
+	return memcmp(out + 16, "\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10", 16) == 0;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+	memset(cracked, 0, sizeof(cracked[0]) * cracked_count);
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
+		unsigned char master[MIN_KEYS_PER_CRYPT][32];
+		int i;
+#ifdef SIMD_COEF_32
+		int lens[MIN_KEYS_PER_CRYPT];
+		unsigned char *pin[MIN_KEYS_PER_CRYPT], *pout[MIN_KEYS_PER_CRYPT];
+
+		for (i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
+			lens[i] = strlen(saved_key[index+i]);
+			pin[i] = (unsigned char*)saved_key[index+i];
+			pout[i] = master[i];
+		}
+		pbkdf2_sha256_sse((const unsigned char**)pin, lens, cur_salt->salt, cur_salt->salt_length, cur_salt->iterations, pout, 32, 0);
+#else
+		for (i = 0; i < MIN_KEYS_PER_CRYPT; ++i)
+			pbkdf2_sha256((unsigned char *)saved_key[index+i], strlen(saved_key[index+i]), cur_salt->salt, cur_salt->salt_length, cur_salt->iterations, master[i], 32, 0);
+#endif
+		for (i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
+			cracked[index+i] = bitwarden_decrypt(cur_salt, master[i]);
```

Comment:
Avoid unnecessary false sharing here by replacing this line with:

```c
if (bitwarden_decrypt(cur_salt, master[i]))
        cracked[index+i] = 1;
```

Better yet, also introduce the `any_cracked` logic to avoid needing to zeroize the entire array on each `crypt_all()` call.